### PR TITLE
Wojtasskorcz wip local snapshot store regex optimization

### PR DESCRIPTION
**Pull request by ktoso**
Follow up with polishing to https://github.com/akka/akka/issues/20821

### CODE REVIEW

File: akka-persistence/src/main/scala/akka/persistence/snapshot/local/LocalSnapshotStore.scala [hunk 1]
```diff
@@ -126,9 +127,20 @@ private[persistence] class LocalSnapshotStore extends SnapshotStore with ActorLo
...
-      case FilenamePattern(pid, snr, tms) ⇒ SnapshotMetadata(URLDecoder.decode(pid, UTF_8), snr.toLong, tms.toLong)
-    }.filter(md ⇒ criteria.matches(md) && !saving.contains(md)).toVector
+    else {
+      // optimised, since noticed to cause slowdowns when many snapshots present:
+      files.iterator.flatMap((file: File) ⇒ {
+        val empty = Iterable.empty[SnapshotMetadata] // since Iterable.empty goes through a new builder, we avoid this
+        extractMetadata(file.getName) match {
+          case Some((pid, snr, tms)) ⇒
+            val md = SnapshotMetadata(URLDecoder.decode(pid, UTF_8), snr, tms)
+            if (criteria.matches(md).&&(!saving.contains(md))) Iterable(md)
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/snapshot/local/LocalSnapshotStore.scala]: strange format of &&

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-persistence/src/main/scala/akka/persistence/snapshot/local/LocalSnapshotStore.scala (hunk 1 above)
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/snapshot/local/LocalSnapshotStore.scala]: Hah, will fix - artifact of me having used "explain scala" to see which conversions were applied and where on the array etc. This rewrite avoids converting Option as Iterable btw (previous one did)

**Comment by ktoso** [comment]
Failure may have been real, investigating

**Comment by ktoso** [comment]
Fixed, my optimisation was wrong, toned down it a bit and brought back the test.

File: akka-persistence/src/main/scala/akka/persistence/snapshot/local/LocalSnapshotStore.scala [hunk 2]
```diff
@@ -161,7 +164,26 @@ private[persistence] class LocalSnapshotStore extends SnapshotStore with ActorLo
...
-  private final class SnapshotFilenameFilter(persistenceId: String) extends FilenameFilter {
+  private def extractMetadata(filename: String): Option[(String, Long, Long)] = {
+    val sequenceNumberEndIdx = filename.lastIndexOf('-')
+    val persistenceIdEndIdx = filename.lastIndexOf('-', sequenceNumberEndIdx - 1)
+    val timestampString = filename.substring(sequenceNumberEndIdx + 1)
+    if (persistenceIdStartIdx >= persistenceIdEndIdx || timestampString.exists(!_.isDigit)) None
+    else {
+      val persistenceId = filename.substring(persistenceIdStartIdx, persistenceIdEndIdx)
+      val sequenceNumber = filename.substring(persistenceIdEndIdx + 1, sequenceNumberEndIdx).toLong
+      val timestamp = filename.substring(sequenceNumberEndIdx + 1).toLong
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/snapshot/local/LocalSnapshotStore.scala]: what happens if `toLong` throws here?

Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/snapshot/local/LocalSnapshotStore.scala]: Addressed now, entire thing is wrapped in a try and we log a warning if unable to extract metadata.

**Comment by ktoso** [comment]
Resolved conflicts and added failure handling in the case of metadata in filename not being a number

**Review by drewhk — APPROVED** [review]

**Comment by ktoso** [comment]
Closing for now.

**Pull request closed without merging**